Source/Urho3D/IO/PackageBuilder.h

```diff
@@ -0,0 +1,85 @@
+//
+// Copyright (c) 2008-2020 the Urho3D project.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#pragma once
+
+namespace Urho3D
+{
+class File;
+class MemoryBuffer;
+
+/// A helper class to create PackageFile.
+class URHO3D_API PackageBuilder
+{
+public:
+    /// Constructor.
+    PackageBuilder();
+
+    /// Build package into memory buffer.
+    bool Create(MemoryBuffer& buffer, bool compress);
+
+    /// Build package into file.
+    bool Create(const SharedPtr<File>& file, bool compress);
```

Comment:
Same thing here. Both `MemoryBuffer` and `File` inherit `AbstractFile`. Could just use `AbstractFile*` here directly and avoid defining multiple methods. Same is true for `Append`.

Reply:
My target was to reflect that PackageBuilder going to keep refcount on file while build the package. I guess I can get away with dynamic_cast